scripts/asr_language_modelling/ngram_lm/eval_beamsearch_ngram.py

```diff
@@ -0,0 +1,186 @@
+            raise ValueError(
+                f"The number of samples in the probabilities file '{args.probs_cache_file}' is not "
+                f"the same as the manifest file. "
+                f"You may need to delete the probabilities cached file."
+            )
+    else:
+        with torch.no_grad():
+            all_logits = asr_model.transcribe(audio_file_paths, batch_size=args.acoustic_batch_size, logprobs=True)
+        all_probs = [softmax(logits) for logits in all_logits]
+        logging.info(f"Writing pickle files of probabilities at {args.probs_cache_file}")
+        with open(args.probs_cache_file, 'wb') as f_dump:
+            pickle.dump(all_probs, f_dump)
+
+    wer_dist_greedy = 0
+    words_count = 0
+    for batch_idx, probs in enumerate(all_probs):
+        preds = np.argmax(probs, axis=1)
+        preds_tensor = torch.tensor(preds, device='cpu').unsqueeze(0)
+        pred_text = asr_model._wer.ctc_decoder_predictions_tensor(preds_tensor)[0]
+        pred_split = pred_text.split()
+        target_split = target_transcripts[batch_idx].split()
+        dist = editdistance.eval(target_split, pred_split)
+        wer_dist_greedy += dist
+        words_count += len(target_split)
+
+    # delete the model to free the memory
+    del asr_model
+
+    if args.decoding_mode == "beamsearch_ngram":
+        if not os.path.exists(args.kenlm_model_file):
+            raise FileNotFoundError(f"Could not find the KenLM model file '{args.kenlm_model_file}'.")
+        lm_path = args.kenlm_model_file
+    else:
+        lm_path = None
+
+    if args.decoding_mode in ["beamsearch_ngram", "beamsearch"]:
+        # creating the beam search decoder
+        beam_search_lm = nemo_asr.modules.BeamSearchDecoderWithLM(
+            vocab=[chr(idx + TOKEN_OFFSET) for idx in range(len(vocabs))],
+            beam_width=args.beam_width,
+            alpha=args.beam_alpha,
+            beta=args.beam_beta,
+            lm_path=lm_path,
+            num_cpus=max(os.cpu_count(), 1),
+            input_tensor=False,
+        )
+
+        wer_dist_best = 0
+        wer_dist_min = 0
+        wer_dist_max = 0
+        sample_idx = 0
+        with open(args.preds_output_file, 'w') as f_out:
+            for batch_idx in tqdm(range(int(np.ceil(len(all_probs) / args.beam_batch_size)))):
+                # disabling type checking
+                with nemo.core.typecheck.disable_checks():
+                    probs_batch = all_probs[batch_idx * args.beam_batch_size : (batch_idx + 1) * args.beam_batch_size]
+                    beams_batch = beam_search_lm.forward(log_probs=probs_batch, log_probs_length=None,)
+
+                for beams_idx, beams in enumerate(beams_batch):
+                    for candidate_idx, candidate in enumerate(beams):
+                        target_split = target_transcripts[sample_idx + beams_idx].split()
+                        pred_text = model_tokenizer.ids_to_text([ord(c) - TOKEN_OFFSET for c in candidate[1]])
```

Comment:
TODO: Add comments regarding this unicode space decoding to get back the original subword id

Reply:
Added